optax/contrib/_sophia_h.py

```diff
@@ -0,0 +1,307 @@
+# Copyright 2023 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Sophia-H optimizer.
+
+A contributed implementation of the Sophia-H optimizer from "Sophia: A Scalable
+Stochastic Second-order Optimizer for Language Model Pre-training"
+(https://arxiv.org/abs/2305.14342) by Hong Liu, Zhiyuan Li, David Hall,
+Percy Liang, and Tengyu Ma.
+
+This contribution is heavily based on the implementation of Sophia by levanter
+(https://github.com/stanford-crfm/levanter) with some changes.
+"""
+from typing import Any, NamedTuple, Optional, Union, Callable
+
+import jax
+from jax import numpy as jnp
+from jax.random import PRNGKey
+
+from optax._src import base
+from optax._src import transform
+from optax._src.numerics import safe_int32_increment
+from optax._src.utils import canonicalize_dtype
+from optax._src.combine import chain
+from optax import tree_utils as otu
+
+
+class SophiaHState(NamedTuple):
+  """State for Sophia-H and similar."""
+
+  count: jax.Array  # shape=(), dtype=jnp.int32
+  mu: base.Updates  # momentum
+  nu: base.Updates  # EMA of hessian
+  key: PRNGKey
+
+
+def scale_by_sophia_h(
+    b1: float = 0.965,
+    b2: float = 0.99,
+    eps: float = 1e-8,
+    gamma: float = 0.01,
+    clip_threshold: Optional[float] = 1.0,
+    update_interval: int = 10,
+    mu_dtype: Optional[Any] = None,
+    pmap_axis_name: Optional[str] = None,
+    seed: PRNGKey = PRNGKey(0),
+    print_win_rate_every_n_steps: int = 0,
+) -> base.GradientTransformationExtraArgs:
+  """Sophia optimizer with hutchinson's estimator for the hessian diagonal.
+
+  The loss function must be passed into sophia's update function to calculate
+  the hessian diagonal. It must accept `params` as its only argument and return
+  only a scalar (the loss).
+
+  For example, considering loss function `loss_fn(params, batch) -> loss, aux`
+  that takes multiple arguments and returns multiple outputs, we must modify it
+  to `loss_fn(params) -> loss` like below:
+
+  `sophia_obj_fn = lambda params: loss_fn(params, batch)[0]`
+
+  Then it can be passed to sophia's update function:
+
+  `updates, state = sophia.update(updates, state, params, obj_fn=sophia_obj_fn)`
+
+  Notes:
+    - TODO filter non-differentiable inputs to jvp
```

Comment:
What do you mean by that?
Are you referring to the ability to take additional input arguments of the function as done in e.g. optax.scale_by_backtracking_linesearch?

Reply:
This is referring to how levanter's version uses equinox.filter_grad but I think we can forget about this for now unless optax wants to host a version of filter_grad similar to equinox's https://github.com/stanford-crfm/levanter/blob/fe414e2fc9116eb7a8bec4f639d86b1bf10be988/src/levanter/optim.py#L67

I think for now I can just delete this line

Reply:
Yes, maybe simpler for now to delete this. Thanks for the answer!